astropy/table/typing.py

```diff
@@ -0,0 +1,15 @@
+from typing import Protocol
+
+from astropy.utils.data_info import MixinInfo
+
+__all__ = [
+    "MixinColumn",
+]
+
+
+class MixinColumn(Protocol):
+    info: MixinInfo
+    shape: tuple[int, ...]
+
+    def __getitem__(self, key): ...
```

Comment:
`MixinColumn` is clearly a [generic type](https://typing.readthedocs.io/en/latest/spec/generics.html#generics) where the parameter must agree with the return type of `__getitem__()`.

Reply:
makes sense, though I'm not sure how well this mixes with a `Protocol`, but let's give it a try !

Reply:
The return type of `__getitem__` can be either `MixinColumn` or a scalar element of `MixinColumn`. I'm not sure how to capture that.

Reply:
I think this was adressed with `TypeVar` + `Protocol[T]` + `@overload`ing. Good to go, @eerovaher ?

Reply:
Good enough for me, but its the sub-package maintainer that should do the final review.